vaadin-multi-select-list-mixin.html

```diff
@@ -0,0 +1,99 @@
+<!--
+@license
+Copyright (c) 2019 Vaadin Ltd.
+This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+-->
+<link rel="import" href="vaadin-list-mixin.html">
+<script>
+  window.Vaadin = window.Vaadin || {};
+
+  /**
+   * A mixin for `nav` elements, facilitating multiple selection of childNodes.
+   *
+   * @polymerMixin
+   */
+  Vaadin.MultiSelectListMixin = superClass => class VaadinMultiSelectListMixin extends Vaadin.ListMixin(superClass) {
+    static get properties() {
+      return {
+        /**
+         * Specifies that multiple options can be selected at once.
+         */
+        multiple: {
+          type: Boolean,
+          value: false,
+          reflectToAttribute: true,
+          observer: '_multipleChanged'
+        },
+
+        /**
+         * Array of indexes of the items selected in the items array
+         */
+        selectedValues: {
+          type: Array,
+          notify: true,
+          value: function() {
+            return [];
+          }
+        }
+      };
+    }
+
+    static get observers() {
+      return [
+        `_enhanceMultipleItems(items, multiple, selected, selectedValues, selectedValues.*)`
+      ];
+    }
+
+    ready() {
+      // Should be attached before click listener in list-mixin
+      this.addEventListener('click', e => this._onMultipleClick(e));
+
+      super.ready();
+    }
+
+    _enhanceMultipleItems(items, multiple, selected, selectedValues) {
+      if (!items || !multiple) {
+        return;
+      }
+
+      if (selectedValues) {
+        const selectedItems = selectedValues.map(selectedId => items[selectedId]);
+        items.forEach(item => item.selected = selectedItems.indexOf(item) !== -1);
+      }
+
+      const lastSelectedItem = this.selectedValues.slice(-1);
+      if (lastSelectedItem && !lastSelectedItem.disabled) {
+        this._scrollToItem(lastSelectedItem);
+      }
+    }
+
+    _onMultipleClick(event) {
+      const item = this._filterItems(event.composedPath())[0];
+      const idx = item && !item.disabled ? this.items.indexOf(item) : -1;
+      if (idx < 0 || !this.multiple) {
+        return;
+      }
+
+      event.preventDefault();
+      if (this.selectedValues.indexOf(idx) !== -1) {
+        this.splice('selectedValues', this.selectedValues.indexOf(idx), 1);
```

Comment:
Clicks fire duplicate `selected-values-changed` events: one for `splices` and another for `length`.

Reply:
Aligned with functionality in `paper-listbox` and i.e. `grid`

Reply:
Yeah, it seems that other components fire duplicate events on array-mutations as well, so I guess that's fine since Polymer works that way.

But one reason why we'd maybe like to assign a new array instead of mutating the existing one is the Flow implementation: `AbstractSinglePropertyField` uses Flow `PropertyChangeListener` internally, and that is fired only when the property value is changed. When the same array is mutated, Flow thinks that the value didn't change and it doesn't fire a value change event at the server side.

Reply:
`<vaadin-checkbox-group>` assigns a new array instead of mutating. I think we should follow that one:
https://github.com/vaadin/vaadin-checkbox/blob/1a06468cc2bd4ba314c4c3fe8923ff74fa85a123/src/vaadin-checkbox-group.html#L245-L253

Reply:
And we should add tests for the `selected-values-changed` event since it clearly is not trivial and obvious how it works.

Reply:
Updated and added test.